browser/ui/sidebar/sidebar_controller.cc

```diff
@@ -243,56 +257,69 @@ void SidebarController::OnTabStripModelChanged(
-        // panels if we're not opening a tab-specific panel.
-        if (!panel_ui->GetCurrentEntryId()) {
-          // Open/Close by tab change doesn't need animation.
-          // UI could refer this for prevent animation.
-          operation_from_active_tab_change_ = true;
-        }
-
-        ActivatePanelItem(wanted_panel.value());
-      }
-    } else {
-      // There is no tab-specific panel for the new active tab
-      if (browser_active_panel_type_.has_value()) {
-        // - restore previous active index
-        ActivatePanelItem(browser_active_panel_type_.value());
-      } else {
-        // - close any tab-specific panel from the previous tab
-        auto current_entry_id = panel_ui->GetCurrentEntryId();
-        // When extension side panel is opened, it's not set to
-        // |browser_active_panel_type_|. We don't call ActivateItemAt() for
-        // extension item as our SidebarModel can't manage it now. Only
-        // tab-specific panel should be closed here if it's used.
-        if (current_entry_id &&
-            (current_entry_id != SidePanelEntryId::kExtension)) {
-          // Open/Close by tab change doesn't need animation.
-          // UI could refer this for preven animation.
-          operation_from_active_tab_change_ = true;
-          panel_ui->Close();
-        }
+  DVLOG(1) << __func__ << " : Active tab changed.";
+
+  // Is there a tab-specific panel for the new active tab?
+  absl::optional<SidebarItem::BuiltInItemType> wanted_panel =
+      selection.new_contents
+          ? SidebarTabHelper::FromWebContents(selection.new_contents)
+                ->active_panel()
+          : absl::nullopt;
+  if (wanted_panel.has_value()) {
+    auto wanted_index = sidebar_model_->GetIndexOf(wanted_panel.value());
+    // built-in items can be removed in-between tab activations, so we might
+    // not get a valid index for this type.
+    if (wanted_index.has_value() &&
+        sidebar_model_->active_index() != wanted_index.value()) {
+      // Don't pass through null values, as we don't want to close existing
+      // panels if we're not opening a tab-specific panel.
+      if (!panel_ui->GetCurrentEntryId()) {
+        // Open/Close by tab change doesn't need animation.
+        // UI could refer this for prevent animation.
+        operation_from_active_tab_change_ = true;
       }
+
+      DVLOG(1) << __func__ << " : Show per-tab panel";
+      ActivatePanelItem(wanted_panel.value());
+    }
+    return;
+  }
+
+  // If active tab has contextual side panel, we should not activate another
+  // panel here. As SidePanelRegistry is only available from views, we need to
+  // ask it to view layer(sidebar_).
+  // TODO(simonhong): Check kChatUI could be contextual entry instead of
+  // handling via SidebarTabHelper.
+  if (sidebar_->HasActiveContextualEntry()) {
+    DVLOG(1) << __func__
+             << " : Just return cause there is active contextual entry.";
+    return;
+  }
```

Comment:
I'm not sure why we need to check this. From my scan, OnEntryShown() writes the entry id, then the code below could handle it properly without this check?

Reply:
I think https://github.com/brave/brave-core/pull/19175#issuecomment-1635273862 could give answer.
W/o this, `ExtensionOpenSidePanelBrowserTest.OpenSidePanel_WindowId_DoesNotOverrideInactiveContextualPanel` failed because proper contextual panel is not set after active tab is changed.